netebpfext/user/nmr_impl.h

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include <condition_variable>
+#include <functional>
+#include <map>
+#include <mutex>
+#include <optional>
+#include <vector>
+
+#include <ebpf_platform.h>
+#include <kernel_um.h>
+#include <netiodef.h>
+#include <../km/netioddk.h>
+
+typedef class _nmr
+{
+  public:
+    typedef void* nmr_provider_handle;
+    typedef void* nmr_client_handle;
+    typedef void* nmr_binding_handle;
+
+    _nmr() = default;
+    ~_nmr() = default;
+
+    /**
+     * @brief Register a provider.
+     *
+     * @param[in] characteristics Characteristics of the provider.
+     * @param[in] context Context passed to the provider.
+     * @return nmr_handle
+     */
+    nmr_provider_handle
+    register_provider(const NPI_PROVIDER_CHARACTERISTICS& characteristics, _In_opt_ void* context);
+
+    /**
+     * @brief Deregister a provider.
+     *
+     * @param[in] provider_handle Handle to the provider.
+     * @return true Caller needs to wait for the deregistering to complete.
+     * @return false Dergistering is complete.
+     */
+    bool
+    deregister_provider(nmr_provider_handle provider_handle);
+
+    /**
+     * @brief Wait for a deregistering to complete.
+     *
+     * @param[in] provider_handle Handle to the provider.
+     */
+    void
+    wait_for_deregister_provider(nmr_provider_handle provider_handle);
+
+    /**
+     * @brief Register a client.
+     *
+     * @param[in] characteristics Characteristics of the client.
+     * @param[in] context Context passed to the client.
+     * @return nmr_handle
+     */
+    nmr_client_handle
+    register_client(const NPI_CLIENT_CHARACTERISTICS& characteristics, _In_ void* context);
+
+    /**
+     * @brief Dergister a client.
+     *
+     * @param[in] client_handle Handle to the client.
+     * @return true Caller needs to wait for the deregistering to complete.
+     * @return false Dergistering is complete.
+     */
+    bool
+    deregister_client(nmr_client_handle client_handle);
+
+    /**
+     * @brief Wait for a deregistering to complete.
+     *
+     * @param[in] client_handle Handle to the client.
+     */
+    void
+    wait_for_deregister_client(nmr_client_handle client_handle);
+
+    /**
+     * @brief Signal that a detach is complete.
+     *
+     * @param[in] binding_handle NMR binding handle.
+     */
+    void
+    binding_detach_complete(nmr_client_handle binding_handle);
+
+    /**
+     * @brief Callback from the client to complete an attach.
+     *
+     * @param[in] binding_handle Binding handle passed to the client.
+     * @param[in] client_binding_context Client's per binding context.
+     * @param[in] client_dispatch Client's dispatch table.
+     * @param[out] provider_binding_context Provider's per binding context.
+     * @param[out] provider_dispatch Provider's dispatch table.
+     * @return NTSTATUS return from ProviderAttachClient.
```

Comment:
Would be helpful, though not required, if this could list specific status codes with retval